src/components/Checkbox/index.tsx

```diff
@@ -1,67 +1,93 @@
 import React, { HTMLAttributes } from 'react';
-import { v4 } from 'uuid';
 import cc from 'classcat';
+import * as CheckboxUi from '@radix-ui/react-checkbox';
+import * as Label from '@radix-ui/react-label';
 
 import CheckEmptyIcon from 'icons/react/CheckEmptyIcon';
 import CheckFilledIcon from 'icons/react/CheckFilledIcon';
 import CheckIndeterminateIcon from 'icons/react/CheckIndeterminateIcon';
 
-export interface CheckboxProps extends HTMLAttributes<HTMLInputElement> {
+type CheckboxExtended = Omit<HTMLAttributes<HTMLInputElement>, 'onChange' | 'ref'>;
+export interface CheckboxProps extends CheckboxExtended {
   /** Label used for a11y attributes _and_ the rendered `label` element */
   label: string;
 
   /** If the supplied `label` prop should be rendered to the screen. */
   showLabel?: boolean;
 
-  /** Ref for input element */
-  inputRef?: React.RefObject<HTMLInputElement>;
-
   /** Sets type `indeterminate` to `true` */
   indeterminate?: boolean;
 
   /** Disables form field when `true` */
   disabled?: boolean;
 
-  /** Supplies an `id` to the input, often used with `showLabel={false}` to connect the label. */
+  /** Supplies an `id` to the input, it is often used with `showLabel={false}` to connect the label. */
   id?: string;
 
   name?: string;
+
+  /*** callback that passes back the checkbox state - true, false or "indeterminate" */
+  onChange?: (checkedEvent: CheckboxUi.CheckedState) => void;
+
+  /*** Value of checkbox */
+  checked?: boolean;
 }
 
 const Checkbox = ({
   showLabel = true,
   indeterminate = false,
   disabled = false,
   label,
-  inputRef,
-  id = v4(),
+  defaultChecked = false,
+  checked,
+  id = undefined,
+  onChange = () => {},
   ...rest
 }: CheckboxProps): JSX.Element => {
   const IconUnchecked = CheckEmptyIcon;
   const IconChecked = indeterminate ? CheckIndeterminateIcon : CheckFilledIcon;
 
+  const [internalChecked, setChecked] = React.useState(checked ?? defaultChecked);
+
+  const handleChange = (checkedEvent: CheckboxUi.CheckedState) => {
+    if (checked === undefined) {
+      setChecked(!internalChecked);
```

Comment:
if the new state depends on the previous internal state it is better to use the callback provided by the setChecked function to make sure that subsequence changes are not overwritten. https://medium.com/geekculture/using-prevstate-with-react-basic-and-complex-f7f8ab4ce1b6

`setChecked((prevInternalChecked) => !prevInternalChecked)`

Reply:
Nice, thanks for the article explanation!